pkg/segment/writer/segstore.go

```diff
@@ -1037,50 +1036,72 @@ func (ss *SegStore) flushBloomIndex(cname string, bi *BloomIndex) uint64 {
 		return 0
 	}
 
-	defer bffd.Close()
-
-	var buf bytes.Buffer
-	bufWriter := bufio.NewWriter(&buf)
-
 	// there is no accurate way to find exactly how many bytes the write.to is going to write
 	// and we need that number , so that we write it first and then the actual bloom data
 	// hence this messiness to write it to some buffer, get the bytesWritten count and then do
 	// the actual write
-	bytesWritten, bferr := bi.Bf.WriteTo(bufWriter)
-	if bferr != nil {
-		log.Errorf("flushBloomIndex: write buf failed fname=%v, err=%v", fname, bferr)
-		return 0
-	}
-
-	bytesWritten += utils.LEN_BLKNUM_CMI_SIZE // for blkNum
-	bytesWritten += 1                         // reserver for CMI_Type
+	bytesWritten := uint32(0)
 
-	// copy the size of blockBloom in uint32
-	if _, err = bffd.Write(toputils.Uint32ToBytesLittleEndian(uint32(bytesWritten))); err != nil {
-		log.Errorf("flushBloomIndex: bloomsize write failed fname=%v, err=%v", fname, err)
+	_, err = bffd.Write([]byte{0, 0, 0, 0})
+	if err != nil {
+		log.Errorf("flushBloomIndex: failed to skip bytes for bloom size fname=%v, err=%v", fname, err)
 		return 0
 	}
+	bytesWritten += 4
 
 	// copy the blockNum
 	if _, err = bffd.Write(toputils.Uint16ToBytesLittleEndian(ss.numBlocks)); err != nil {
-		log.Errorf("flushBloomIndex: bloomsize write failed fname=%v, err=%v", fname, err)
+		log.Errorf("flushBloomIndex: block num write failed fname=%v, err=%v", fname, err)
 		return 0
 	}
+	bytesWritten += utils.LEN_BLKNUM_CMI_SIZE
 
 	// write CMI type
 	if _, err = bffd.Write(utils.CMI_BLOOM_INDEX); err != nil {
 		log.Errorf("flushBloomIndex: CMI Type write failed fname=%v, err=%v", fname, err)
 		return 0
 	}
+	bytesWritten += 1
 
 	// write the blockBloom
-	_, bferr = bi.Bf.WriteTo(bffd)
-	if bferr != nil {
-		log.Errorf("flushBloomIndex: write blockbloom failed fname=%v, err=%v", fname, bferr)
+	bloomSize, err := bi.Bf.WriteTo(bffd)
+	if err != nil {
+		log.Errorf("flushBloomIndex: write blockbloom failed fname=%v, err=%v", fname, err)
+		return 0
+	}
+	bytesWritten += uint32(bloomSize)
+
+	err = bffd.Close()
```

Comment:
I noticed that the file is opened at the beginning, but there are multiple return points before it's closed. This could lead to the file not being closed properly if an error occurs,

I would suggest refactoring the code that opens and writes to the file into a separate function. In this way, you can do `defer` immediately like you have done when you have opened the file for the second time.

Reply:
yeah sorry, I missed it, I have addressed it in the latest commit.